elasticsearch/security_api.go

```diff
@@ -0,0 +1,220 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package elasticsearch
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strconv"
+)
+
+// requires manage_security cluster privilege
+func CreateAPIKey(client Client, apikeyReq CreateApiKeyRequest) (CreateApiKeyResponse, error) {
+	response := client.JSONRequest(http.MethodPut, "/_security/api_key", apikeyReq)
```

Comment:
This comment is sort of done 3 times, so I'll reply just this one and close the others: what do you mean with "avoid running the rest of this function"?: on the very first line of the very next function call, the error is checked.

All this is doing is to push down error handling to keep higher level functions as lean as possible. This way, new functionality is trivial to implement: just define request and response types (which you would have to do anyways), and pass whatever endpoint to `JSONRequest`.
Otherwise, error handling would need to be duplicated on every single function.

TL;DR Code reuse

Reply:
Sorry, missed this response previously.  Running the reset of that function does a heap allocation that could be avoided if the error was checked immediately.  ince apm server won't be creating keys all the time the extra allocation is fine too.  I do think it's usually best to check errors close to where they occur to avoid issues where later refactors may not realize the caller's behavior - at least a comment would be useful in this case.